### PR TITLE
Quick Army Purhase in Town Screen

### PR DESCRIPTION
**Pull request by FeniksFire**
Mouse click on castle icon (it is now clickable) - quick recruit of all
available creatures.

It's not ideal like a version from hd mod, don't have graphics from there and don't have time to finish all features now, but should be enough for normal player.

### CODE REVIEW

File: client/windows/CCastleInterface.cpp [hunk 1]
```diff
@@ -990,7 +996,8 @@ void CCastleInterface::recreateIcons()
 
 	hall = new CTownInfo( 80, 413, town, true);
 	fort = new CTownInfo(122, 413, town, false);
-
+	fastArmyPurhase = std::make_shared<CButton>(Point(122, 413), "itmcl.def", CButton::tooltip("quick army purhase"), [&](){builds->enterToTheQuickRecruitmentWindow();});
```
Comment (alexvins) [inline client/windows/CCastleInterface.cpp]: Please do not use smart pointers to `CIntObject` descendants when DISPOSE capture flag set (OBJ_CONSTRUCTION_CAPTURING_ALL macro set all flags)

File: client/windows/QuickRecruitmentWindow.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+			i->slider->setAmount(i->maxamount);
+		i->slider->moveTo(i->slider->getValue());
+	}
+	totalCost->createItems(LOCPLINT->cb->getResourceAmount() - allAvailableResources);
+	totalCost->set(LOCPLINT->cb->getResourceAmount() - allAvailableResources);
+}
+
+QuickRecruitmentWindow::QuickRecruitmentWindow(const CGTownInstance * townd) : CWindowObject(PLAYER_COLORED | BORDERED), town(townd)
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL;
```
Comment (alexvins) [inline client/windows/QuickRecruitmentWindow.cpp]: Same issue with smart pointers mixed with DISPOSE flag. If you want to use smart pointers you need modified `OBJ_CONSTRUCTION_CAPTURING_ALL` macro: `CIntObject::DISPOSE` bit must be cleared in defActions.

**Comment by FeniksFire** [comment]
@alexvins I think it should be fine now. Not the best way but since we are using macros there I have added one more to them with the ability to set action flag. I think it shouldn't be too distracting if someone would like to refactor this part.

File: client/windows/CCastleInterface.cpp [hunk 3]
```diff
@@ -990,7 +996,8 @@ void CCastleInterface::recreateIcons()
 
 	hall = new CTownInfo( 80, 413, town, true);
 	fort = new CTownInfo(122, 413, town, false);
-
+	fastArmyPurhase = new CButton(Point(122, 413), "itmcl.def", CButton::tooltip("quick army purhase"), [&](){builds->enterToTheQuickRecruitmentWindow();});
```
Comment (alexvins) [inline client/windows/CCastleInterface.cpp]: `"quick army purhase"` string should be put to `config/translate.json`

File: client/windows/QuickRecruitmentWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			cards.push_back(std::make_shared<CreaturePurhaseCard>(VLC->creh->creatures[crid], pos, town->creatures[i].first, this));
+			pos.x += 108;
+		}
+	}
+	totalCost = std::make_shared<CreatureCostBox>(Rect((this->pos.w/2)-45, pos.y + 260, 97, 74), "");
+}
+
+void QuickRecruitmentWindow::initWindow()
+{
+	pos.x = 238;
```
Comment (alexvins) [inline client/windows/QuickRecruitmentWindow.cpp]: Position (at least horizontal) should be relative to castle interface.

File: client/windows/QuickRecruitmentWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+void QuickRecruitmentWindow::setCreaturePurhaseCards()
+{
+	Point pos = Point(8,64);
+	for (int i=0; i < GameConstants::CREATURES_PER_TOWN; i++)
+	{
+		if(!town->town->creatures.at(i).empty() && !town->creatures.at(i).second.empty() && town->creatures[i].first)
+		{
+			CreatureID crid = town->creatures[i].second[0];
+			if(town->creatures[i].second.size() > 1)
+				crid = town->creatures[i].second[1];
```
Comment (alexvins) [inline client/windows/QuickRecruitmentWindow.cpp]: Mods can add more than two creatures at level. (Alternative creatures are officially supported)

File: client/windows/QuickRecruitmentWindow.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+void QuickRecruitmentWindow::setMaxButton()
+{
+	maxButton = std::make_shared<CButton>(Point((pos.w/2)-113, 417), "IRCBTNS.DEF", CButton::tooltip(), [&](){ maxAllCards(cards); });
+	maxButton->setImageOrder(0, 1, 2, 3);
+}
+
+void QuickRecruitmentWindow::setCreaturePurhaseCards()
+{
+	Point pos = Point(8,64);
```
Comment (alexvins) [inline client/windows/QuickRecruitmentWindow.cpp]: With only one creature available this looks ugly. Card is aligned to left, and too much empty space after it.

File: client/windows/QuickRecruitmentWindow.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+			i->slider->setAmount(i->slider->getValue() + maxAmount);
+		else
+			i->slider->setAmount(i->maxamount);
+		i->slider->moveTo(i->slider->getValue());
+	}
+	totalCost->createItems(LOCPLINT->cb->getResourceAmount() - allAvailableResources);
+	totalCost->set(LOCPLINT->cb->getResourceAmount() - allAvailableResources);
+}
+
+QuickRecruitmentWindow::QuickRecruitmentWindow(const CGTownInstance * townd) : CWindowObject(PLAYER_COLORED | BORDERED), town(townd)
```
Comment (alexvins) [inline client/windows/QuickRecruitmentWindow.cpp]: Put initializers on next line

**Comment by FeniksFire** [comment]
@alexvins fast fix.

**Comment by FeniksFire** [comment]
@alexvins fast fix.

**Comment by FeniksFire** [comment]
omg, I close it by mistake ;d

**Pull request merged**